File: web/src/services/api/habits.ts
import { http } from "./client";
import { ENDPOINTS } from "./endpoints";
import type { Task } from "./tasks";
import type { UUID } from "@/types/primitive";
import type { ListResponse } from "@/types/pagination";

// Types for habits
export interface Habit {
  id: UUID;
  title: string;
  description?: string | null;
  start_date: string;
  duration_days: number;
  cadence_frequency?: string | null;
  cadence_weekdays?: string[] | null;
  cadence_monthdays?: number[] | null;
  target_per_cycle?: number | null;
  status: string;
  stats?: HabitStats | null;
  task_id?: UUID | null;
  task?: Task | null;
}

export interface HabitCreate {
  title: string;
  description?: string;
  start_date: string;
  duration_days?: number;
  end_date?: string | null;
  repeat_count?: number | null;
  cadence_frequency?: string | null;
  cadence_weekdays?: string[] | null;
  cadence_monthdays?: number[] | null;
  target_per_cycle?: number | null;
  task_id?: UUID | null;
}

export interface HabitUpdate {
  title?: string;
  description?: string;
  start_date?: string;
  duration_days?: number;
  end_date?: string | null;
  repeat_count?: number | null;
  cadence_frequency?: string | null;
  cadence_weekdays?: string[] | null;
  cadence_monthdays?: number[] | null;
  target_per_cycle?: number | null;
  status?: string;
  task_id?: UUID | null;
}

export interface HabitAction {
  id: UUID;
  habit_id: UUID;
  action_date: string;
  status: string;
  notes?: string | null;
}

export interface HabitActionHabitSummary {
  title: string;
  description?: string | null;
  start_date: string;
  duration_days: number;
}

export interface HabitActionUpdate {
  status?: string;
  notes?: string;
}

export interface HabitStats {
  habit_id: UUID;
  total_actions: number;
  completed_actions: number;
  missed_actions: number;
  skipped_actions: number;
  progress_percentage: number;
  current_streak: number;
  longest_streak: number;
}

interface HabitOverview {
  habit: Habit;
  stats: HabitStats;
}

export type HabitOverviewResponse = HabitOverview;

interface HabitListMeta {
  status_filter?: string | null;
}

interface HabitActionListMeta {
  status_filter?: string | null;
  center_date?: string | null;
  days_before?: number | null;
  days_after?: number | null;
}

interface HabitActionByDateListMeta {
  action_date?: string | null;
}

interface HabitActionRangeListMeta {
  start_date?: string | null;
  end_date?: string | null;
  reference_date?: string | null;
}

export type HabitOverviewListResponse = ListResponse<
  HabitOverview,
  HabitListMeta
>;

export type HabitListResponse = ListResponse<Habit, HabitListMeta>;

export type HabitActionListResponse = ListResponse<
  HabitAction,
  HabitActionListMeta
>;

export interface HabitActionWithHabit extends HabitAction {
  habit: HabitActionHabitSummary;
}

export type HabitActionByDateListResponse = ListResponse<
  HabitActionWithHabit,
  HabitActionByDateListMeta
>;

export type HabitActionRangeListResponse = ListResponse<
  HabitActionWithHabit,
  HabitActionRangeListMeta
>;

export interface HabitTaskAssociationsResponse {
  associations: Record<UUID, Habit[]>;
}

export interface HabitActionsQueryParams {
  page?: number;
  size?: number;
  statusFilter?: string;
  centerDate?: string;
  daysBefore?: number;
  daysAfter?: number;
}

export const habitsApi = {
  async getAll(
    statusFilter?: string,
    params?: { page?: number; size?: number },
  ): Promise<HabitListResponse> {
    return http.get<HabitListResponse>(ENDPOINTS.HABITS.BASE, {
      status_filter: statusFilter,
      page: params?.page,
      size: params?.size,
    });
  },

  async getOverviews(
    statusFilter?: string,
    params?: { page?: number; size?: number },
  ): Promise<HabitOverviewListResponse> {
    return http.get<HabitOverviewListResponse>(ENDPOINTS.HABITS.OVERVIEWS, {
      status_filter: statusFilter,
      page: params?.page,
      size: params?.size,
    });
  },

  async getActionsByDate(date: string): Promise<HabitActionByDateListResponse> {
    return http.get<HabitActionByDateListResponse>(
      ENDPOINTS.HABITS.ACTIONS_BY_DATE(date),
    );
  },

  async getActionsInRange(params: {
    startDate: string;
    endDate: string;
    referenceDate: string;
    page?: number;
    size?: number;
  }): Promise<HabitActionRangeListResponse> {
    return http.get<HabitActionRangeListResponse>(
      ENDPOINTS.HABITS.ACTIONS_IN_RANGE,
      {
        start_date: params.startDate,
        end_date: params.endDate,
        reference_date: params.referenceDate,
        page: params.page,
        size: params.size,
      },
    );
  },

  async getById(id: UUID): Promise<Habit> {
    return http.get<Habit>(ENDPOINTS.HABITS.BY_ID(id));
  },

  async getOverview(id: UUID): Promise<HabitOverviewResponse> {
    return http.get<HabitOverviewResponse>(ENDPOINTS.HABITS.OVERVIEW_BY_ID(id));
  },

  async create(habit: HabitCreate): Promise<Habit> {
    return http.post<Habit>(ENDPOINTS.HABITS.BASE, habit);
  },

  async update(id: UUID, habit: HabitUpdate): Promise<Habit> {
    return http.patch<Habit>(ENDPOINTS.HABITS.BY_ID(id), habit);
  },

  async delete(id: UUID): Promise<void> {
    return http.delete<void>(ENDPOINTS.HABITS.BY_ID(id));
  },

  async getActions(
    habitId: UUID,
    {
      page,
      size,
      statusFilter,
      centerDate,
      daysBefore,
      daysAfter,
    }: HabitActionsQueryParams = {},
  ): Promise<HabitActionListResponse> {
    return http.get<HabitActionListResponse>(
      ENDPOINTS.HABITS.ACTIONS(habitId),
      {
        page,
        size,
        status_filter: statusFilter,
        center_date: centerDate,
        days_before: daysBefore,
        days_after: daysAfter,
      },
    );
  },

  async updateAction(
    habitId: UUID,
    actionId: UUID,
    actionUpdate: HabitActionUpdate,
  ): Promise<HabitAction> {
    return http.patch<HabitAction>(
      ENDPOINTS.HABITS.ACTION_BY_ID(habitId, actionId),
      actionUpdate,
    );
  },

  async getHabitTaskAssociations(): Promise<HabitTaskAssociationsResponse> {
    return http.get<HabitTaskAssociationsResponse>(
      ENDPOINTS.HABITS.TASK_ASSOCIATIONS,
    );
  },
};

File: src/lifeos_web/routers/habits.py
"""Habit action endpoints used by the local planning UI."""

from __future__ import annotations

from datetime import date
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession

from lifeos_cli.db.models.habit import Habit
from lifeos_cli.db.models.habit_action import HabitAction
from lifeos_cli.db.services import habit_actions as habit_action_services
from lifeos_cli.db.services import habits as habit_services
from lifeos_web.deps import get_db_session
from lifeos_web.schemas import HabitActionUpdate, HabitCreate, HabitUpdate, ListResponse, Pagination
from lifeos_web.serialization import to_jsonable

router = APIRouter(prefix="/habits", tags=["habits"])
SessionDep = Annotated[AsyncSession, Depends(get_db_session)]


def _habit_model_payload(habit: Habit) -> dict[str, object]:
    """Serialize a Habit SQLAlchemy model for the local Web API."""
    return {
        "id": str(habit.id),
        "title": habit.title,
        "description": habit.description,
        "start_date": habit.start_date.isoformat(),
        "duration_days": habit.duration_days,
        "cadence_frequency": habit.cadence_frequency,
        "cadence_weekdays": habit.cadence_weekdays,
        "cadence_monthdays": getattr(habit, "cadence_monthdays", None),
        "target_per_cycle": habit.target_per_cycle,
        "status": habit.status,
        "task_id": str(habit.task_id) if habit.task_id else None,
    }


async def _habit_payload(session: AsyncSession, habit_id: UUID) -> dict[str, object]:
    habit = await habit_services.get_habit(session, habit_id=habit_id)
    if habit is None:
        raise HTTPException(status_code=404, detail=f"Habit {habit_id} was not found")
    return _habit_model_payload(habit)


def _habit_action_payload(action: object) -> dict[str, object]:
    """Serialize a habit action occurrence for Web consumers."""
    if isinstance(action, HabitAction):
        payload: dict[str, object] = {
            "id": str(action.id),
            "habit_id": str(action.habit_id),
            "action_date": action.action_date.isoformat(),
            "status": action.status,
            "notes": getattr(action, "notes", None),
        }
    else:
        jsonable = to_jsonable(action)
        assert isinstance(jsonable, dict)
        payload = jsonable
        payload.pop("habit_title", None)
        payload.pop("created_at", None)
        payload.pop("updated_at", None)
        payload.pop("deleted_at", None)
    return payload


async def _action_with_habit_summary(
    session: AsyncSession,
    action: object,
) -> dict[str, object]:
    payload = _habit_action_payload(action)
    habit = await habit_services.get_habit(session, habit_id=UUID(str(payload["habit_id"])))
    if habit is None:
        raise HTTPException(status_code=404, detail=f"Habit {payload['habit_id']} was not found")
    payload["habit"] = {
        "title": habit.title,
        "description": habit.description,
        "start_date": habit.start_date.isoformat(),
        "duration_days": habit.duration_days,
    }
    return payload


def _zero_stats(habit_id: UUID) -> dict[str, object]:
    return {
        "habit_id": str(habit_id),
        "total_actions": 0,
        "completed_actions": 0,
        "missed_actions": 0,
        "skipped_actions": 0,
        "progress_percentage": 0,
        "current_streak": 0,
        "longest_streak": 0,
    }


@router.get("/", response_model=ListResponse)
async def list_habits(
    session: SessionDep,
    page: int = 1,
    size: int = 100,
    status_filter: str | None = None,
) -> ListResponse:
    """List habits for the local Web UI."""
    rows = await habit_services.list_habits(
        session,
        status=status_filter,
        limit=size,
        offset=(page - 1) * size,
    )
    items = [_habit_model_payload(row) for row in rows]
    return ListResponse(
        items=items,
        pagination=Pagination(page=page, size=size, total=len(items), pages=1 if items else 0),
        meta={"status_filter": status_filter},
    )


@router.get("/overviews", response_model=ListResponse)
async def list_habit_overviews(
    session: SessionDep,
    page: int = 1,
    size: int = 100,
    status_filter: str | None = None,
) -> ListResponse:
    """List habits with frontend-compatible zeroed stats."""
    habits = await habit_services.list_habits(
        session,
        status=status_filter,
        limit=size,
        offset=(page - 1) * size,
    )
    items = [
        {
            "habit": _habit_model_payload(habit),
            "stats": _zero_stats(habit.id),
        }
        for habit in habits
    ]
    return ListResponse(
        items=items,
        pagination=Pagination(page=page, size=size, total=len(items), pages=1 if items else 0),
        meta={"status_filter": status_filter},
    )


@router.get("/habit-task-associations/")
async def get_habit_task_associations(session: SessionDep) -> dict[str, object]:
    """Return habits grouped by associated task for frontend planning components."""
    habits = await habit_services.list_habits(session, limit=500, offset=0)
    associations: dict[str, list[dict[str, object]]] = {}
    for habit in habits:
        if habit.task_id is None:
            continue
        task_id = str(habit.task_id)
        associations.setdefault(task_id, []).append(_habit_model_payload(habit))
    return {"associations": associations}


@router.get("/{habit_id}/overview")
async def get_habit_overview(habit_id: UUID, session: SessionDep) -> dict[str, object]:
    """Load one habit overview with frontend-compatible stats."""
    habit = await _habit_payload(session, habit_id)
    return {"habit": habit, "stats": _zero_stats(habit_id)}


@router.post("/")
async def create_habit(
    payload: HabitCreate,
    session: SessionDep,
) -> dict[str, object]:
    """Create a habit."""
    try:
        habit = await habit_services.create_habit(
            session,
            title=payload.title,
            description=payload.description,
            start_date=payload.start_date,
            duration_days=payload.duration_days,
            end_date=payload.end_date,
            repeat_count=payload.repeat_count,
            cadence_frequency=payload.cadence_frequency,
            cadence_weekdays=payload.cadence_weekdays,
            cadence_monthdays=payload.cadence_monthdays,
            target_per_cycle=payload.target_per_cycle,
            task_id=payload.task_id,
        )
    except (LookupError, ValueError) as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _habit_model_payload(habit)


@router.patch("/{habit_id}")
async def update_habit(
    habit_id: UUID,
    payload: HabitUpdate,
    session: SessionDep,
) -> dict[str, object]:
    """Update a habit."""
    fields = payload.model_fields_set
    clear_description = "description" in fields and payload.description is None
    clear_weekdays = "cadence_weekdays" in fields and payload.cadence_weekdays is None
    clear_monthdays = "cadence_monthdays" in fields and payload.cadence_monthdays is None
    clear_task = "task_id" in fields and payload.task_id is None
    try:
        habit = await habit_services.update_habit(
            session,
            habit_id=habit_id,
            title=payload.title,
            description=payload.description,
            clear_description=clear_description,
            start_date=payload.start_date,
            duration_days=payload.duration_days,
            end_date=payload.end_date,
            repeat_count=payload.repeat_count,
            cadence_frequency=payload.cadence_frequency,
            cadence_weekdays=payload.cadence_weekdays,
            clear_weekdays=clear_weekdays,
            cadence_monthdays=payload.cadence_monthdays,
            clear_monthdays=clear_monthdays,
            target_per_cycle=payload.target_per_cycle,
            status=payload.status,
            task_id=payload.task_id,
            clear_task=clear_task,
        )
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _habit_model_payload(habit)


@router.delete("/{habit_id}", status_code=204)
async def delete_habit(
    habit_id: UUID,
    session: SessionDep,
) -> None:
    """Soft-delete a habit."""
    try:
        await habit_services.delete_habit(session, habit_id=habit_id)
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.get("/actions/by-date/{action_date}")
async def list_actions_by_date(
    action_date: date,
    session: SessionDep,
) -> ListResponse:
    """List materialized habit actions for one local date."""
    actions = await habit_action_services.list_habit_actions(
        session,
        date_values=(action_date,),
        limit=500,
    )
    items = [await _action_with_habit_summary(session, action) for action in actions]
    return ListResponse(
        items=items,
        pagination=Pagination(page=1, size=500, total=len(items), pages=1 if items else 0),
        meta={"action_date": action_date.isoformat()},
    )


@router.get("/actions")
async def list_actions_in_range(
    session: SessionDep,
    start_date: Annotated[date, Query()],
    end_date: Annotated[date, Query()],
    reference_date: Annotated[date | None, Query()] = None,
    page: Annotated[int, Query(ge=1)] = 1,
    size: Annotated[int, Query(ge=1, le=1000)] = 1000,
) -> ListResponse:
    """List materialized habit actions for one local planning date range."""
    try:
        total = await habit_action_services.count_habit_actions(
            session,
            start_date=start_date,
            end_date=end_date,
            reference_date=reference_date,
        )
        actions = await habit_action_services.list_habit_actions(
            session,
            start_date=start_date,
            end_date=end_date,
            reference_date=reference_date,
            limit=size,
            offset=(page - 1) * size,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    items = [await _action_with_habit_summary(session, action) for action in actions]
    pages = (total + size - 1) // size if total else 0
    return ListResponse(
        items=items,
        pagination=Pagination(page=page, size=size, total=total, pages=pages),
        meta={
            "start_date": start_date.isoformat(),
            "end_date": end_date.isoformat(),
            "reference_date": reference_date.isoformat() if reference_date else None,
        },
    )


@router.get("/{habit_id}/actions")
async def list_actions_for_habit(
    habit_id: UUID,
    session: SessionDep,
    start_date: Annotated[date | None, Query()] = None,
    end_date: Annotated[date | None, Query()] = None,
    size: Annotated[int, Query(ge=1, le=500)] = 100,
) -> ListResponse:
    """List materialized/scheduled actions for one habit."""
    try:
        actions = await habit_action_services.list_habit_actions(
            session,
            habit_id=habit_id,
            start_date=start_date,
            end_date=end_date,
            limit=size,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    items = [_habit_action_payload(action) for action in actions]
    return ListResponse(
        items=items,
        pagination=Pagination(page=1, size=size, total=len(items), pages=1 if items else 0),
        meta={
            "status_filter": None,
            "center_date": None,
            "days_before": None,
            "days_after": None,
        },
    )


@router.patch("/{habit_id}/actions/{action_id}")
async def update_action(
    habit_id: UUID,
    action_id: UUID,
    payload: HabitActionUpdate,
    session: SessionDep,
) -> dict[str, object]:
    """Update one habit action."""
    del habit_id
    fields = payload.model_fields_set
    clear_notes = "notes" in fields and payload.notes is None
    try:
        action = await habit_action_services.update_habit_action(
            session,
            action_id=action_id,
            status=payload.status,
            notes=payload.notes,
            clear_notes=clear_notes,
        )
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _habit_action_payload(action)


@router.patch("/{habit_id}/actions/by-date/{action_date}")
async def update_action_by_date(
    habit_id: UUID,
    action_date: date,
    payload: HabitActionUpdate,
    session: SessionDep,
) -> dict[str, object]:
    """Update one habit action by habit/date, materializing it when needed."""
    fields = payload.model_fields_set
    clear_notes = "notes" in fields and payload.notes is None
    try:
        action = await habit_action_services.update_habit_action_by_date(
            session,
            habit_id=habit_id,
            action_date=action_date,
            status=payload.status,
            notes=payload.notes,
            clear_notes=clear_notes,
        )
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return _habit_action_payload(action)
